refactor: Drop noop gcc version checks

hebasto opened the pull request:
Since #20413 the minimum required GCC version is 7.

hebasto committed:
Drop noop gcc version checks

Since #20413 the minimum required GCC version is 7.

Co-authored-by: practicalswift <practicalswift@users.noreply.github.com>

fanquake commented:
I think these are included in #19183. Also, unless we added that check, I'm not sure if we want to modify nanobench.

hebasto commented:
> I think these are included in #19183. Also, unless we added that check, I'm not sure if we want to modify nanobench.
> 
> Edit: Saying that. #19183 is being split up/possibly blocked, so if the nanobench change is removed, I don't mind if this is merged earlier.

Removed nanobench change, and added credits to @practicalswift.

MarcoFalke commented on src/test/fuzz/multiplication_overflow.cpp:
```diff
@@ -14,7 +14,7 @@
 #define HAVE_BUILTIN_MUL_OVERFLOW
 #endif
-#elif defined(__GNUC__) && (__GNUC__ >= 5)
+#elif defined(__GNUC__)
```
isn't this covered by the previous check already?

https://gcc.gnu.org/onlinedocs/gcc-7.1.0/gcc/Integer-Overflow-Builtins.html

hebasto replied:
Thanks! Updated.

hebasto replied:
Reverted back due to @fanquake's [comment](https://github.com/bitcoin/bitcoin/pull/20491#issuecomment-734135165).

laanwj commented:
Code review ACK a526ab4fca37d8148e45705c97da49da3aa36fae

fanquake commented:
I'm not sure this is correct, or at-least, now that this is only using `__has_builtin()` for the builtin checks, they wont be defined for GCC < 10, as that was the first version to get [`__has_builtin()`](https://gcc.gnu.org/gcc-10/changes.html).
> New built-in functions: 
> The __has_builtin built-in preprocessor operator can be used to query support for built-in functions provided by GCC and other compilers that support it.

Is that the desired behavior? I assume that's why it also has a check for `__GNUC__`.

i.e:

```cpp
#include <iostream>

int main() {
#if defined(__has_builtin)
        std::cout << "has __has_builtin()\n";
#endif
        return 0;
}
```

```bash
g++ --version
g++ (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0

g++ test.cpp && ./a.out
# not avaliable

g++-10 test.cpp && ./a.out
has __has_builtin()
```

hebasto commented:
> I'm not sure this is correct, or at-least, now that this is only using `__has_builtin()` for the builtin checks, they wont be defined for GCC < 10, as that was the first version to get [`__has_builtin()`](https://gcc.gnu.org/gcc-10/changes.html).

Updated ([reverted back](https://github.com/bitcoin/bitcoin/pull/20491#discussion_r532107913)).

fanquake reviewed (approved):
ACK 830ddf413934226d0b6ca99165916790cc52ca18
